### PR TITLE
Add Support for Wren Keyboard

### PR DESCRIPTION
**Pull request by walterhanley**
## Description

This pull request adds support for the Wren, an asymmetric split keyboard. Includes two layouts, one bare-bones (default), and one more fleshed-out (walterhanley). More details about the project can be found on the [project page](https://github.com/walterhanley/wren-keyboard). 

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [X] Keyboard (addition or update)
- [X] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Add code for my [Wren keyboard](https://github.com/walterhanley/wren-keyboard)
* Added default and my keymap

## Checklist

- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/wren/keymaps/walterhanley/keymap.c [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+enum layer_names {
+    _BASE,
+    _SYMBOL,
+    _NAV
+};
+
+// Defines the keycodes used by our macros in process_record_user
+enum custom_keycodes {
+    NSEAL = SAFE_RANGE,
+    QMKURL
```
Comment (zvecr) [inline keyboards/wren/keymaps/walterhanley/keymap.c:33]:
```suggestion
```

File: keyboards/wren/keymaps/walterhanley/keymap.c [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            }
+            break;
+        case QMKURL:
+            if (record->event.pressed) {
+                // when keycode QMKURL is pressed
+                SEND_STRING("https://qmk.fm/\n");
+            } else {
+                // when keycode QMKURL is released
+            }
+            break;
```
Comment (zvecr) [inline keyboards/wren/keymaps/walterhanley/keymap.c:99]:
```suggestion
```

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/wren/keymaps/default/keymap.c [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        }
+    } else if (index == 1) { /* Second encoder */
+        if (clockwise) {
+            tap_code(KC_VOLU);
+        } else {
+            tap_code(KC_VOLD);
+        }
+    }
+    return false;
+}
```
Comment (drashna) [inline keyboards/wren/keymaps/default/keymap.c:53]:
```suggestion
```

File: keyboards/wren/keymaps/walterhanley/keymap.c [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        }
+    } else if (index == 1) { /* Second encoder */
+        if (clockwise) {
+            tap_code(KC_VOLU);
+        } else {
+            tap_code(KC_VOLD);
+        }
+    }
+    return false;
+}
```
Comment (drashna) [inline keyboards/wren/keymaps/walterhanley/keymap.c:79]:
```suggestion
```

File: keyboards/wren/wren.c [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "wren.h"
```
Comment (drashna) [inline keyboards/wren/wren.c:17]:
```suggestion
#include "wren.h"

bool encoder_update_kb(uint8_t index, bool clockwise) {
    if (!encoder_update_user(index, clockwise)) { return false; }
    if (index == 0) { /* First encoder */
        if (clockwise) {
            tap_code(KC_TAB);
        } else {
            tap_code(KC_TAB);
        }
    } else if (index == 1) { /* Second encoder */
        if (clockwise) {
            tap_code(KC_VOLU);
        } else {
            tap_code(KC_VOLD);
        }
    }
    return true;
}

```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/wren/wren.c (hunk 5 above)
Comment (walterhanley) [reply keyboards/wren/wren.c:17]: Happy to make this change, but out of curiosity is there a reason why this is preferred? If users want to change the default encoder keycode assignments, should they change them in keyboard.c or overwrite them with a new function in their keymap.c?

**Review by drashna — APPROVED** [review]

**Review by noroadsleft — APPROVED** [review]

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**